swarm/swap/swap.go

```diff
@@ -0,0 +1,589 @@
+	payAt                = big.NewInt(-4096 * 10000)   // threshold that triggers payment {request} (bytes)
+	dropAt               = big.NewInt(-4096 * 12000)   // threshold that triggers disconnect (bytes)
+
+	ErrInsufficientFunds = errors.New("Insufficient funds")
+	ErrNotAccountedMsg   = errors.New("Message does not need accounting")
+)
+
+const (
+	chequebookDeployRetries = 5
+	chequebookDeployDelay   = 1 * time.Second // delay between retries
+)
+
+// SwAP Swarm Accounting Protocol with
+//      Swift Automatic  Payments
+// a peer to peer micropayment system
+type Swap struct {
+	chequeManager *ChequeManager
+	stateStore    state.Store
+	lock          sync.RWMutex
+	peers         map[discover.NodeID]*SwapPeer
+	local         *Params // local peer's swap parameters
+}
+
+type SwapPeer struct {
+	*protocols.Peer
+	lock        sync.RWMutex
+	swapAccount *Swap
+	handlerFunc func(context.Context, interface{}) error
+	balance     *big.Int
+	storeID     string
+}
+
+type EntryDirection bool
+
+const (
+	DebitEntry  EntryDirection = true
+	CreditEntry EntryDirection = false
+)
+
+type SwapAccountedMsgType interface {
+	GetMsgPrice() *big.Int
+}
+
+//Handler for received messages
+func (sp *SwapPeer) RunAccountedProtocol(protocolHandler func(ctx context.Context, msg interface{}) error) error {
+	//the `peer.Run` function is a loop, so in order to pre-/post-process a message with accounting,
+	//we need to save the actual handler
+	sp.handlerFunc = protocolHandler
+
+	//then run the handler loop function
+	return sp.Run(sp.handleAccountedMsg)
+}
+
+//get a peer's balance
+func (swap *Swap) GetPeerBalance(peer discover.NodeID) *big.Int {
+	if p, ok := swap.peers[peer]; ok {
+		return p.balance
+	}
+	return nil
+}
+
+//Handle a received message; this is the handler loop function.
+//Check if it needs accounting, and if yes, apply accounting logic:
+//Check for sufficient funds, perform operation, then account
+func (sp *SwapPeer) handleAccountedMsg(ctx context.Context, msg interface{}) error {
+	var err error
+	var price *big.Int
+
+	//the message is one which needs accounting...
+	if _, ok := msg.(SwapAccountedMsgType); ok {
+		//..so first check if there are enough funds for the operation available
+		//(for crediting, this means if we are not essentially "overdrafting", or crossing the threshold)
+		price, err = sp.checkAvailableFunds(ctx, msg, CreditEntry)
+		//if not (or some other error occured), return error
+		if err != nil {
+			//also, if the error is indeed insufficient funds, then disconnect the peer
+			if err == ErrInsufficientFunds {
+				log.Error("Insufficient funds, dropping peer")
+				sp.Drop(err)
+			}
+			return err
+		}
+		//at this point we know there are sufficient funds, so process the message
+		err = sp.handlerFunc(ctx, msg)
+		if err == nil {
+			//and if no errors occurred, finally book the entry
+			sp.AccountMsgForPeer(ctx, msg, price, CreditEntry)
```

Comment:
not sure why we need two accounting calls here. check available funds should do the accounting.
irrespective of the actual handle calls success

Reply:
Questionable. My approach was:
1. Check for available funds
2. Process actual message
3. Do accounting

If we do accounting in 1. what happens if there is a failure in 2.? We have to rollback.

However, we can account **after** processing the message, when we are sure that we actually charging for something which really happened.

The problem with this approach however is, as I documented in the gitter channel, is that we check for available funds in 1. , then process the message, **but in the meantime the balance may have changed**, as the process is not transactional. Correct would be to check again for available funds when doing accounting, which takes us back to the same problem....Another possibility is to actually `lock` on the `Swap` object while we are processing the message, but then we are indeed blocking, impacting performance. But these would be the correct procedures of financial accounting from my experience.

Reply:
do handle, if successful or not, then just do the accounting based on the msg.
do send, if successful, then just do the accounting based on the msg.
surely the tally must be locked, parallellizable
see any issues with this?

Reply:
The issue is that we need to check for available funds **before** we do the accounting. We can not just handle or send, then account, as, if there are not enough funds, the accounting would fail.